torch/_inductor/optimize_indexing.py

```diff
@@ -0,0 +1,473 @@
+
+    @classmethod
+    def wrap(cls, arg):
+        if isinstance(arg, ValueRanges):
+            return arg
+        assert isinstance(arg, (int, float, bool))
+        return ValueRanges(arg, arg)
+
+    @classmethod
+    def unary_map(cls, x, fn):
+        "map lower and upper bound with fn"
+        x = cls.wrap(x)
+        return ValueRanges(fn(x.lower), fn(x.upper))
+
+    @classmethod
+    def checked_unary_map(cls, x, fn):
+        "check the max and min of computed upper and lower bound for the output"
+        out = cls.unary_map(x, fn)
+        return ValueRanges(min(out.lower, out.upper), max(out.lower, out.upper))
+
+    @classmethod
+    def binary_map(cls, x, y, fn):
+        "map upper and lower bounds accessing corresponding values of inputs"
+        x, y = cls.wrap(x), cls.wrap(y)
+
+        return ValueRanges(
+            fn(x.lower, y.lower),
+            fn(x.upper, y.upper),
+        )
+
+    @classmethod
+    def binary_map_products(cls, a, b, fn):
+        "compute the product of all lower and upper bounds and take min and max"
+        a, b = cls.wrap(a), cls.wrap(b)
+        products = [
+            fn(x, y)
+            for x, y in itertools.product([a.lower, a.upper], [b.lower, b.upper])
+        ]
+        return ValueRanges(min(products), max(products))
+
+
+class ValueRangeAnalysis(object):
+    def __init__(self):
+        boolean_operators = (
+            "eq",
+            "ne",
+            "lt",
+            "gt",
+            "le",
+            "ge",
+            "and_",
+            "or_",
+            "xor",
+            "logical_and",
+            "logical_or",
+            "logical_not",
+        )
+        for op in boolean_operators:
+            setattr(self, op, self.bool_handler)
+
+    @staticmethod
+    def bool_handler(*args, **kwargs):
+        # just assuming bools can have both values
+        return ValueRanges(
+            sympy.logic.boolalg.BooleanFalse, sympy.logic.boolalg.BooleanTrue
+        )
+
+    def load(self, name: str, index: sympy.Expr):
+        return ValueRanges(-math.inf, math.inf)
+
+    def store(self, name, index, value, mode=None):
+        return
+
+    def reduction(self, name, dtype, src_dtype, reduction_type, index, value):
+        return ValueRanges(-math.inf, math.inf)
+
+    def index_expr(self, index, dtype):
+        assert isinstance(index, ValueRanges)
+        return index
+
+    @staticmethod
+    def to_dtype(x, dtype: torch.dtype):
+        return ValueRanges.wrap(x)
+
+    @staticmethod
+    def constant(value, dtype):
+        # using nan makes subsequent computation throw, and for the purposes of optimization
+        # returning -math.inf - math.inf is equivalent to giving up
+        if math.isnan(value):
+            return ValueRanges(-math.inf, math.inf)
+        if isinstance(value, int):
+            return ValueRanges(
+                sympy.core.numbers.Integer(value), sympy.core.numbers.Integer(value)
```

Comment:
```suggestion
                sympy.Integer(value), sympy.Integer(value)
```

We had some issues with this in fbcode.